packages/discussions/lib/setDiscussionPreferences.js

```diff
@@ -61,10 +61,22 @@ module.exports = async ({
   const existingDP = await transaction.public.discussionPreferences.findOne(findQuery)
   const user = await transaction.public.users.findOne({ id: userId })
 
+  let anonymousDiff = null
+  if (anonymity) {
+    if (!existingDP || existingDP.anonymousDifferentiator === null) {
+      const lastUsed = await transaction.public.discussionPreferences.findOne({
+        discussionId: discussion.id,
+        'anonymousDifferentiator !=': null
+      }, { orderBy: { anonymousDifferentiator: 'DESC' } })
+      anonymousDiff = lastUsed ? lastUsed.anonymousDifferentiator + 1 : 1
+    }
+  }
+
```

Comment:
I would recommend to move this logic to `submitComment`. By doing so we could ensure that the `anonymousDifferentiator` is only assigned once some one actually participates. Additionally we could also handle anonymity enforced debates by creating a `discussionPreference` with `anonymous` and  `anonymousDifferentiator` on submit.

Reply:
I see where you are coming from. Unfortunatly, this could lead to some edge-cases that will need to be faught with more code.

Say `Ada Lovelace` has already commented in a thread. If she now changes her anonymity settings *a posteriori* without submitting a new comment, the `anonymousDifferentiator` will not be set.

I'll try to figure out a clean way.

Reply:
Ah true that. Then I would recommend leaving this here but only setting `anonymousDifferentiator` when:

- it is called from `submitComment` with a new special arg `setAnonymousDifferentiator` where we'd now always call it (potential optimise to avoid unnecessary mutations when a `anonymousDifferentiator` is not needed)
- or when not called from `submitComment` / with `setAnonymousDifferentiator` but already has a comment (`comments.find({userId, discussionId: discussion.id })`)